src/digraph.rs

```diff
@@ -2183,6 +2195,79 @@ impl PyDiGraph {
         }
     }
 
+    /// Create a new :class:`~retworkx.PyDiGraph` object from an adjacency matrix
+    /// with matrix elements of type ``complex``
+    ///
+    /// This method can be used to construct a new :class:`~retworkx.PyDiGraph`
+    /// object from an input adjacency matrix. The node weights will be the
+    /// index from the matrix. The edge weights will be a float value of the
+    /// value from the matrix.
+    ///
+    /// This differs from the
+    /// :meth:`~retworkx.PyDiGraph.from_adjacency_matrix` in that the type of
+    /// the elements of the input matrix in this method must be a ``complex``
+    /// (specifically a ``numpy.complex128``) and the output graph edge weights
+    /// will be ``complex`` too. While in
+    /// :meth:`~retworkx.PyDiGraph.from_adjacency_matrix` the matrix elements
+    /// are of type ``float`` (specifically ``numpy.float64``) and the edge
+    /// weights in the output graph will be ``float`` too.
+    ///
+    /// :param ndarray matrix: The input numpy array adjacency matrix to create
+    ///     a new :class:`~retworkx.PyDiGraph` object from. It must be a 2
+    ///     dimensional array and be a ``complex``/``np.complex128`` data type.
+    /// :param float null_value: An optional float that will treated as a null
+    ///     value. If any element in the input matrix is this value it will be
+    ///     treated as not an edge. By default this is ``0.0``
+    ///
+    /// :returns: A new graph object generated from the adjacency matrix
+    /// :rtype: PyDiGraph
+    #[staticmethod]
+    #[pyo3(text_signature = "(matrix, /)")]
+    pub fn from_complex_adjacency_matrix<'p>(
+        py: Python<'p>,
+        matrix: PyReadonlyArray2<'p, Complex64>,
+        null_value: Option<Complex64>,
+    ) -> PyDiGraph {
```

Comment:
I don't actually know if this will work as expected. Mostly because the `PyReadonlyArray2` is talking to the numpy C api to access the array in memory directly which will have the dtype set and we're basically accessing a raw c type instead of accessing it via python. While you can set an [`object_`](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.object_) as the dtype in a numpy array that's an explicit choice. I'll test this out locally though and confirm either way and report back here

Reply:
Indeed it looks like [because of  rust-numpy implementation](https://pyo3.github.io/rust-numpy/numpy/array/struct.PyArray.html#data-type-and-dimension), we won't be able to support `PyObject` or `numpy.object_` as they will not implement [numpy::TypeNum](https://pyo3.github.io/rust-numpy/numpy/types/trait.TypeNum.html)